lektor_tags.py

```diff
@@ -148,3 +166,103 @@ def get_all_tags(self, parent):
 
     def ignore_missing(self):
         return bool_from_string(self.get_config().get("ignore_missing"), False)
+
+    def _get_tagcount(self):
+        """Return a dictionary of tag counts.
+
+        Compute it only once, and return the "cached" version then.
+        """
+        if self._tagcount is not None:
+            return self._tagcount
+
+        # Count tags, to be aggregated as "tag weights". Note that tags that
+        # only appear in non-discoverable pages are ignored.
+        self._tagcount = collections.defaultdict(int)
+        for page in get_ctx().pad.query(self.get_parent_path()):
+            for tag in page[self.get_tag_field_name()]:
+                self._tagcount[tag] += 1
+        return self._tagcount
+
+    def _get_minmax_count(self):
+        """Return a tuple of minimum and maximum tag counts."""
+        tagcount = self._get_tagcount()
+        if tagcount:
+            return min(tagcount.values()), max(tagcount.values())
+        else:
+            # There isn't a single tag
+            return 0, 0
+
+    def _weight_count(self):
+        """Map each tag to the number of pages tagged with it."""
+        return self._get_tagcount()
+
+    def _weight_linear(self, a, b):
+        """Map each tag with a number between `a` and `b`.
+
+        The less used tag is mapped `a`, the most used tag is mapped `b`.
+        Mapping is done using a linear function.
+        """
+        tagcount = self._get_tagcount()
+        mincount, maxcount = self._get_minmax_count()
+        if mincount == maxcount:
+            return {tag: a for tag in tagcount}
+        return {
+            tag: (
+                ((b - a) * count + a * maxcount - b * mincount) / (maxcount - mincount)
+            )
+            for tag, count in tagcount.items()
+        }
+
+    def _weight_lineargroup(self, groups):
+        """Map each tag with an item of list `groups`.
+
+        The less used tag is mapped with the first item, the most used tag is mapped with the last item.
+        Mapping is done using a linear function.
+        """
+        weights = self._weight_linear(0, len(groups) - 1)
+        return {tag: groups[int(round(weights[tag]))] for tag in self._get_tagcount()}
+
+    def _weight_log(self, a, b):
```

Comment:
I know it's a bit more verbose (and less mathematical :sweat_smile:) but I think lower/upper_bound or just lower/upper are simpler to understand when looking at the signature.

If you want to you can still assign a = lower_bound when you use them.

Reply:
Done in 53508fe751882a5efe4bfb06570be1e6133cb8b7